File: backend/app/api/v1/endpoints/driver_desk_sim.py
from fastapi import APIRouter, HTTPException

from app.communication.driver_desk_test_rig import driver_desk_test_rig
from app.schemas.driver_desk_sim import (
    DriverDeskSimInputUpdateRequest,
    DriverDeskSimPulseRequest,
    DriverDeskSimStartRequest,
)

router = APIRouter()


@router.get("/status", summary="Get driver-desk protocol simulator status")
def get_driver_desk_sim_status() -> dict:
    return driver_desk_test_rig.snapshot()


@router.post("/start", summary="Start local driver-desk protocol simulator and backend bridge")
def start_driver_desk_sim(command: DriverDeskSimStartRequest) -> dict:
    try:
        return driver_desk_test_rig.start(
            vehicle_id=command.vehicle_id,
            simulator_host=command.simulator_host,
            simulator_port=command.simulator_port,
            pulse_width_ms=command.pulse_width_ms,
        )
    except OSError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/stop", summary="Stop local driver-desk protocol simulator and backend bridge")
def stop_driver_desk_sim() -> dict:
    return driver_desk_test_rig.stop()


@router.patch("/input", summary="Update sustained driver-desk simulator inputs")
def update_driver_desk_sim_input(command: DriverDeskSimInputUpdateRequest) -> dict:
    try:
        return driver_desk_test_rig.update_input(command.updates)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/pulse", summary="Pulse one momentary driver-desk simulator input")
def pulse_driver_desk_sim_input(command: DriverDeskSimPulseRequest) -> dict:
    try:
        return driver_desk_test_rig.pulse(command.field_name, command.duration_ms)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: frontend/src/api/driverDeskSim.js
import http from '@/api/index'

export function getDriverDeskSimStatus() {
  return http.get('/driver-desk-sim/status', { suppressErrorLog: true })
}

export function startDriverDeskSim(payload) {
  return http.post('/driver-desk-sim/start', payload)
}

export function stopDriverDeskSim() {
  return http.post('/driver-desk-sim/stop', {})
}

export function updateDriverDeskSimInput(updates) {
  return http.patch('/driver-desk-sim/input', { updates })
}

export function pulseDriverDeskSim(fieldName, durationMs = null) {
  return http.post('/driver-desk-sim/pulse', {
    field_name: fieldName,
    ...(durationMs != null ? { duration_ms: durationMs } : {}),
  })
}
